src/backend/utils/readPixelsHelper.ts

```diff
@@ -33,6 +33,12 @@ export class ReadPixelsHelper {
         // Empty error list.
         gl.getError();
 
+        // hack to get WebXR layer inputs working from Quest 2.
+        // The framebuffers have type UNSIGNED_NORMALIZED, which needs to be converted into something that is a valid format for readPixel.
+        if (type === WebGlConstants.UNSIGNED_NORMALIZED.value) {
```

Comment:
Are we sure this does not break anything else ? could it be conditioned to smthg else if it has unwilling potential side effects ?

Reply:
@sebavan - okay, I think I have a better understanding of what's happening now.

The code path that brings us to the `readPixels` call from a WebXR layers experience is from `VisualState.readFrameBufferAttachmentFromRenderBuffer`, which also calls `VisualState.getCapture` and finally `ReadPixelsHelper.readPixels`.

The `type` that eventually gets passed through to `readPixelsHelper.readPixels` there is the `componentType` calculated earlier, which is the `FRAMEBUFFER_ATTACHMENT_COMPONENT_TYPE` for the color attachment of the framebuffer. That can be `FLOAT`, `INT`, `UNSIGNED_INT`, `SIGNED_NORMALIZED`, or `UNSIGNED_NORMALIZED`.

However, the `readPixels` call requires a `type` that is the pixel data type. Of those, `FLOAT`, `UNSIGNED_INT`, and `INT` are already valid, but the others need additional information (e.g. stride, format) to know what the proper pixel data type is from the component type. It so happened that the hack here is correct for Quest 2, or produces results that look correct, but may not be so in other situations. (I'm not too familiar with this - I found [this code from Chromium](https://chromium.googlesource.com/angle/angle/+/refs/heads/main/src/libANGLE/validationES.cpp#190) that validates whether a pixel data type is correct for a given component type, but not how to actually get that pixel data type. Notably, according to that code, `FLOAT` is _not_ a valid pixel data type for an `UNSIGNED_NORMALIZED` component, but `UNSIGNED_BYTE` can be.)

This does mean that the code as written is a little odd - because `readPixelsHelper.readPixels` has checks for when the `type` is `UNSIGNED_BYTE`, but a component type will _never_ be `UNSIGNED_BYTE`. I did notice that other calls to `readPixel` will explicitly provide an `UNSIGNED_BYTE` for the type rather than use the component type, but that's not the code path we end up going down here.

Do you have any insight as to why, when capturing screenshots, `getCapture` and then `readPixels` is sometimes called with a component type? Either way, without this change as written, Spector.JS errors out on WebXR layers experiences, because it'd go down the code path later on in the same `readPixels` function to change type to `FLOAT`, which is invalid. I'm happy to make more changes as needed here, but any large-scale changes to accommodate all valid formats is probably out of scope of this PR.

Reply:
Thanks a lot, can we then just remove the quest specific comment ? as it sounds a valid option no mater what ?

Reply:
I modified the comment to be more generic; personally I think it'd be good to make an issue to handle other component types, and link that here, but that's out of scope for this PR.